File: python/modcdp/launcher/RemoteBrowserLauncher.py
# MODCDP_TRANSLATE: KEEP THIS FILE TRANSLATED ACROSS TYPESCRIPT, PYTHON, AND GO.
# Keep all shapes, signatures, behavior, and tests 1:1 in sync with:
# - ./js/src/launcher/RemoteBrowserLauncher.ts
# - ./go/modcdp/launcher/RemoteBrowserLauncher.go
from __future__ import annotations

from typing import cast

from ..launcher.BrowserLauncher import LauncherOptions, BrowserLauncher, LaunchedBrowser, resolveCdpWebSocketUrl


class RemoteBrowserLauncher(BrowserLauncher):
    def launch(self, options: LauncherOptions | None = None) -> LaunchedBrowser:
        merged = {**self.options, **dict(options or {})}
        cdp_url = merged.get("launcher_remote_cdp_url")
        if not cdp_url:
            raise RuntimeError("launcher.launcher_mode=remote requires launcher_remote_cdp_url.")
        # cdp_url is resolved here so downstream transports can dial it directly.
        cdp_url = resolveCdpWebSocketUrl(cast(str, cdp_url), "launcher_remote_cdp_url")
        self.launched = {"cdp_url": cdp_url, "close": lambda: None}
        return self.launched

File: go/modcdp/launcher/RemoteBrowserLauncher.go
// MODCDP_TRANSLATE: KEEP THIS FILE TRANSLATED ACROSS TYPESCRIPT, PYTHON, AND GO.
// Keep all shapes, signatures, behavior, and tests 1:1 in sync with:
// - ./js/src/launcher/RemoteBrowserLauncher.ts
// - ./python/modcdp/launcher/RemoteBrowserLauncher.py
package launcher

import "fmt"

type RemoteBrowserLauncher struct {
	BrowserLauncher
}

func NewRemoteBrowserLauncher(options LaunchOptions) *RemoteBrowserLauncher {
	return &RemoteBrowserLauncher{BrowserLauncher: NewBrowserLauncher(options)}
}

func (l *RemoteBrowserLauncher) Launch(options LaunchOptions) (*LaunchedBrowser, error) {
	cdpURL := firstString(options.LauncherRemoteCDPURL, l.Config.LauncherRemoteCDPURL)
	if cdpURL == "" {
		return nil, fmt.Errorf("launcher.launcher_mode=remote requires launcher_remote_cdp_url")
	}
	resolvedCDPURL, err := websocketURLFor(cdpURL)
	if err != nil {
		return nil, err
	}
	// CDPURL is resolved here so downstream transports can dial it directly.
	launched := &LaunchedBrowser{CDPURL: resolvedCDPURL, Close: func() {}}
	l.Launched = launched
	return launched, nil
}
